packages/grid/x-data-grid/src/hooks/core/useGridStateInitialization.ts

```diff
@@ -36,7 +36,7 @@ export const useGridStateInitialization = <PrivateApi extends GridPrivateApiComm
         newState = state;
       }
 
-      if (apiRef.current.state === newState) {
+      if (isDeepEqual(apiRef.current.state, newState)) {
```

Comment:
At first sight, it looks like a no-go. Why is comparing the state not enough? 

```suggestion
      if (apiRef.current.state === newState) {
```

Actually, it might be great to deep Object.freeze the state to prevent any mutations.

Reply:
> Why is comparing the state not enough?

because states are objects. and `===` does not work on objects. 
what about `JSON.stringify`, its faster.

Reply:
> because states are objects. and === does not work on objects.

`===` works on object, it does a reference comparison.
Why would you need a value comparison on the state here ?

Reply:
> Why would you need a value comparison on the state here ?

I have removed it for now. but 'stateChange' is getting triggered multiple times with the same state.
check this demo: https://codesandbox.io/s/objective-bash-1hfd3s?file=/demo.tsx:1646-1653

Reply:
The solution would probably to find why this happen and add a more specific check on the guilty `setState`